Allow `isolatedModules: true` with `noEmitOnError: true`

## Search Terms

- isolatedModules
- noEmitOnError

## Suggestion

Allow isolatedModules: true with noEmitOnError: true

## Use Cases

### What do you want to use this for?

My understanding of the purpose of `isolatedModules: true` is that it does extra validation to ensure that separate compilation is safe. We want to use it for that purpose.

Allowing emit in the presence of errors is contrary to that goal. We're doing the validation to *stop* dangerous code from getting through.

### What shortcomings exist with current approaches?

We use TypeScript in our compilation pipeline.

The only workaround I can think of is some hack to compile to a different location and copy files over once checks pass. This doesn't seem great for watch mode, and could be error-prone or racy.

## Examples

**tsconfig.json**: 

```json
{
    "compilerOptions": {
        "noEmitOnError": true,
        "isolatedModules": true,
    }
}
```

Actual behavior: `tsconfig.json(5,9): error TS5053: Option 'noEmitOnError' cannot be specified with option 'isolatedModules'.`

Expected behavior: no error message, just works

## Checklist

My suggestion meets these guidelines:

* [x ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -2707,10 +2707,6 @@ namespace ts {
                     createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, getEmitDeclarationOptionName(options), "isolatedModules");
                 }
 
-                if (options.noEmitOnError) {
-                    createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, "noEmitOnError", "isolatedModules");
-                }
-
                 if (options.out) {
                     createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, "out", "isolatedModules");
                 }

--- a/tests/baselines/reference/isolatedModulesNoEmitOnError.errors.txt
+++ b/tests/baselines/reference/isolatedModulesNoEmitOnError.errors.txt
@@ -1,6 +1,7 @@
-error TS5053: Option 'noEmitOnError' cannot be specified with option 'isolatedModules'.
+tests/cases/compiler/file1.ts(1,14): error TS2322: Type '3' is not assignable to type 'string'.
 
 
-!!! error TS5053: Option 'noEmitOnError' cannot be specified with option 'isolatedModules'.
-==== tests/cases/compiler/file1.ts (0 errors) ====
-    export var x;
+==== tests/cases/compiler/file1.ts (1 errors) ====
+    export const x: string = 3;
+                 ~
+!!! error TS2322: Type '3' is not assignable to type 'string'.

--- a/tests/baselines/reference/isolatedModulesNoEmitOnError.symbols
+++ b/tests/baselines/reference/isolatedModulesNoEmitOnError.symbols
@@ -1,4 +1,4 @@
 === tests/cases/compiler/file1.ts ===
-export var x;
->x : Symbol(x, Decl(file1.ts, 0, 10))
+export const x: string = 3;
+>x : Symbol(x, Decl(file1.ts, 0, 12))
 

--- a/tests/baselines/reference/isolatedModulesNoEmitOnError.types
+++ b/tests/baselines/reference/isolatedModulesNoEmitOnError.types
@@ -1,4 +1,5 @@
 === tests/cases/compiler/file1.ts ===
-export var x;
->x : any
+export const x: string = 3;
+>x : string
+>3 : 3
 

--- a/tests/cases/compiler/isolatedModulesNoEmitOnError.ts
+++ b/tests/cases/compiler/isolatedModulesNoEmitOnError.ts
@@ -3,4 +3,4 @@
 // @target: es6
 
 // @filename: file1.ts
-export var x;
+export const x: string = 3;

